### PR TITLE
feat: cli to upload data to s3

### PR DESCRIPTION
**Pull request by Ben-Cheng**
Fixes: #68

### CODE REVIEW

File: topo_processor/aws_sdk/s3_upload.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+from topo_processor.metadata.collection import Collection
+from topo_processor.metadata.item import Item
+from topo_processor.util.checksum import multihash_as_hex_file, multihash_as_hex_stac_object
+from topo_processor.util.time import time_in_ms
+
+
+async def upload_data(item: Item, bucket: str):
+    start_time = time_in_ms()
+    s3 = boto3.client("s3")
```
Comment (Ben-Cheng) [inline topo_processor/aws_sdk/s3_upload.py:15]: dont make one client per upload, should pass it around instead

File: topo_processor/aws_sdk/s3_upload.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+import boto3
+from linz_logger import get_log
+
+from topo_processor.metadata.collection import Collection
+from topo_processor.metadata.item import Item
+from topo_processor.util.checksum import multihash_as_hex_file, multihash_as_hex_stac_object
+from topo_processor.util.time import time_in_ms
+
+
+async def upload_data(item: Item, bucket: str):
```
Comment (Ben-Cheng) [inline topo_processor/aws_sdk/s3_upload.py:13]: accept client as parameter

Comment (MDavidson17) [inline topo_processor/aws_sdk/s3_upload.py:13]: stac output file should have tiff in it so tiff should not be in this function

File: topo_processor/aws_sdk/s3_upload.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    start_time = time_in_ms()
+    s3 = boto3.client("s3")
+    s3.upload_file(
+        Filename=item.path,
+        Bucket=bucket,
+        Key=f"{item.output_filename}.tiff",
+        ExtraArgs={"ContentType": "image/tiff", "Metadata": {"hash": multihash_as_hex_file(item.path)}},
```
Comment (Ben-Cheng) [inline topo_processor/aws_sdk/s3_upload.py:20]: do not rehash the file but get it from the stac

File: topo_processor/aws_sdk/s3_upload.py [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+        Metadata={"hash": multihash_as_hex_stac_object(stac_object_as_bytes(stac_object))},
+        ContentType="application/json",
+    )
+    get_log().debug(
+        "S3 Metadata Object Uploaded",
+        duration=time_in_ms() - start_time,
+        Bucket=bucket,
+        Key=f"{stac_object.output_filename}.json",
+    )
+
```
Comment (Ben-Cheng) [inline topo_processor/aws_sdk/s3_upload.py:47]: write the json to disk and upload it as file

File: topo_processor/aws_sdk/s3_upload.py [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+async def upload(item: Item, bucket: str):
+    await upload_data(item, bucket)
+    await upload_metadata(item, bucket)
+
+
+def stac_object_as_bytes(stac_object: Union[Item, Collection]) -> str:
+    if isinstance(stac_object, Item):
+        return json.dumps(stac_object.stac_item.to_dict(), indent=2).encode("utf-8")
+    return json.dumps(stac_object.stac_collection.to_dict(), indent=2).encode("utf-8")
```
Comment (Ben-Cheng) [inline topo_processor/aws_sdk/s3_upload.py:57]: use `stac.write_file` instead

File: topo_processor/cli/main.py [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+        await upload_to_s3(collection, target)
+    get_log().debug(
+        "Upload Completed",
+        location=target,
+        data_type=collection.data_type.value,
+        duration=time_in_ms() - start_time,
+    )
+
+
+async def upload_to_local_disk(collection, target):
```
Comment (Ben-Cheng) [inline topo_processor/cli/main.py:66]: move this function to a different file

File: topo_processor/cli/main.py [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+        stac.write_file(obj=item.stac_item, include_self_link=True, dest_href=f"{target}/{item.output_filename}.json")
+        copyfile(item.path, f"{target}/{item.output_filename}.tiff")
+    stac.write_file(
+        obj=collection.stac_collection,
+        include_self_link=True,
+        dest_href=f"{target}/{collection.output_filename}.json",
+    )
+
+
+async def upload_to_s3(collection, target):
```
Comment (Ben-Cheng) [inline topo_processor/cli/main.py:77]: move this function to a different file

File: topo_processor/cli/main.py (hunk 6 above)
Comment (Ben-Cheng) [reply topo_processor/cli/main.py:66]: make this one async

File: topo_processor/cli/main_test.py [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+    assert os.path.isfile(os.path.join(target, "C8054", "29659.tiff"))
+    assert os.path.isfile(os.path.join(target, "C8054", "collection.json"))
+
+    with open(os.path.join(target, "C8054", "29659.json")) as item_json_file:
+        item_metadata = json.load(item_json_file)
+    assert (
+        item_metadata["properties"]["checksum:multihash"]
+        == "1220f7ea9017ef596cc650369493dc3993d9ab5c99055bbc5c7bac9d4e061c5f2f7a"
+    )
+    shutil.rmtree(os.path.join(target, "C8054"))
```
Comment (Ben-Cheng) [inline topo_processor/cli/main_test.py:33]: do more checking:
1. collection.json has pointers to items
2. item json has pointer to tiff

Comment (Ben-Cheng) [reply topo_processor/cli/main_test.py:33]: look for a better way to handle tmp dir (e.g. with)

File: topo_processor/metadata/item_test.py [hunk 9]
```diff
@@ -4,8 +4,5 @@
 
 
 def test_output_filenames():
-
-    tiff_path = "dir1/dir2/blahblahblah123.tiff"
-    collection = Collection("title", "description", "license", DataType.ImageryHistoric)
-    item = Item(tiff_path, collection)
-    assert item.output_filename == "blahblahblah123.tiff.json"
+    pass
+    # output file name now set in relevant metadata loader
```
Comment (Ben-Cheng) [inline topo_processor/metadata/item_test.py:8]: remove this test / update the test

File: topo_processor/metadata/loaders/metadata_loader_checksum.py [hunk 10]
```diff
@@ -14,4 +14,4 @@ def is_applicable(self, item: Item) -> bool:
     async def add_metadata(self, item: Item) -> None:
         if "checksum" not in item.stac_item.stac_extensions:
             item.stac_item.stac_extensions.append("checksum")
-        item.stac_item.properties["checksum:multihash"] = multihash_as_hex(item.path)
+        item.stac_item.properties["checksum:multihash"] = multihash_as_hex_file(item.path)
```
Comment (Ben-Cheng) [inline topo_processor/metadata/loaders/metadata_loader_checksum.py:17]: make multihash_as_hex_file async

File: topo_processor/util/checksum.py [hunk 11]
```diff
@@ -1,13 +1,23 @@
...
 CHUNK_SIZE = 1024 * 1024  # 1MB
 
 
-def multihash_as_hex(file_path: str) -> str:
+def multihash_as_hex_file(file_path: str) -> str:
     file_hash = hashlib.sha256()
     with open(file_path, "rb") as file:
         while chunk := file.read(CHUNK_SIZE):
             file_hash.update(chunk)
     return multihash.to_hex_string(multihash.encode(file_hash.digest(), "sha2-256"))
```
Comment (Ben-Cheng) [inline topo_processor/util/checksum.py:17]: this needs to be async

File: topo_processor/util/checksum.py [hunk 12]
```diff
@@ -5,7 +5,7 @@
 CHUNK_SIZE = 1024 * 1024  # 1MB
 
 
-def multihash_as_hex(file_path: str) -> str:
+async def multihash_as_hex(file_path: str) -> str:
```
Comment (blacha) [inline topo_processor/util/checksum.py:8]: by just adding `async` does actually make this async.

you will need to use a async file reader to actually read from the disk in async manner

```
with aiofiles.open(file_path, 'rb' as file:
  while chunk := await file.read(CHUNK_SIZE)
```

File: topo_processor/uploader/s3.py [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+
+async def upload_items(collection: Collection, target: str):
+    to_upload = []
+    for item in collection.items:
+        # TODO save to temp location
+        # for metadata
+        await write_stac_object(item, f"temp/{item.item_output_path}")
+        hash_value = await multihash_as_hex(f"temp/{item.item_output_path}")
+        to_upload.append(
+            upload_file(f"temp/{item.item_output_path}", f"{item.item_output_path}", "application/json", hash_value, target)
```
Comment (blacha) [inline topo_processor/uploader/s3.py:25]: you do not need to use `f` strings if you just have a single variable

```python
f"{item.item_output_path}"
```

File: topo_processor/cli/main_test.py [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+import pytest
+
+from topo_processor.metadata import DataType, create_collection
+
+from .main import upload_to_local_disk
+
+
+@pytest.fixture(autouse=True)
+def teardown():
+    yield
```
Comment (blacha) [inline topo_processor/cli/main_test.py:15]: why the `yeild` if you are going to do something werid, its generally a good idea to comment it as why

File: topo_processor/cli/main_test.py [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+    assert os.path.isfile(os.path.join(target, "C8054", "29659.tiff"))
+    assert os.path.isfile(os.path.join(target, "C8054", "collection.json"))
+
+
+@pytest.mark.asyncio
+async def test_item_contents():
+
+    source = os.path.join(os.getcwd(), "test_data", "tiffs", "C8054")
+    datatype = "imagery.historic"
+    target = gettempdir()
```
Comment (blacha) [inline topo_processor/cli/main_test.py:39]: should use a actual temp folder not the `/tmp` i think there is a `mktempdir` or similar on `tempfile`

**Review by blacha — CHANGES_REQUESTED** [review]

File: topo_processor/uploader/s3.py [hunk 16]
```diff
@@ -0,0 +1,66 @@
...
+        collection.collection_output_path,
+        "application/json",
+        hash_value,
+        target,
+    )
+
+
+async def upload_file(filepath: str, key: str, content_type: str, hash_value: str, bucket: str):
+    start_time = time_in_ms()
+    s3.upload_file(
```
Comment (palmerj) [inline topo_processor/uploader/s3.py:55]: is boto3 upload_file async friendly? i.e it won't block?

Comment (Ben-Cheng) [reply topo_processor/uploader/s3.py:55]: No, please refer to this ticket: https://linzgovt.atlassian.net/jira/software/projects/TDE/boards/581/backlog?selectedIssue=TDE-2

**Review by blacha — APPROVED** [review]

**Pull request merged**